middleware/src/middleware.go

```diff
@@ -335,56 +381,211 @@ func (middleware *Middleware) RestoreHSMSecret() rpcmessages.ErrorResponse {
 }
 
 // UserAuthenticate returns an ErrorResponse struct in response to a rpcserver request.
-// FIXME: This is a dummy implementation of the UserAuthenticate RPC call.
-// in the future this should return an AuthentificationResponse with e.g. an JWT.
-// This currently uses the `dummyIsBaseSetup` boolean, which should be removed when the proper authentication is implemented.
-func (middleware *Middleware) UserAuthenticate(args rpcmessages.UserAuthenticateArgs) rpcmessages.ErrorResponse {
-	// TODO: replace the dummyIsBaseSetup with a proper variable loaded from e.g. redis
-	// dummyIsBaseSetup should only be used for the dummy UserAuthenticate RPC and gets reset on middleware restart
-	if !middleware.dummyIsBaseSetup {
-		if args.Username == "admin" && args.Password == "ICanHasPassword?" {
-			// middleware.dummyIsBaseSetup is only set to true after the initial admin password is changed
-			return rpcmessages.ErrorResponse{Success: true}
-		}
-	} else if middleware.dummyIsBaseSetup {
-		dummyUsers := map[string]string{
-			"admin":   middleware.dummyAdminPassword,
-			"satoshi": "shift1",
-			"dev":     "dev",
-		}
-		if expectedPasssword, userIsInMap := dummyUsers[args.Username]; userIsInMap {
-			if args.Password == expectedPasssword {
-				return rpcmessages.ErrorResponse{Success: true}
+// To check if the user should be authenticated from default values, the bool 'isMiddlewarePasswordSet' is read from redis
+func (middleware *Middleware) UserAuthenticate(args rpcmessages.UserAuthenticateArgs) rpcmessages.UserAuthenticateResponse {
+	//Use this structure to hold the bases's user list in memory
+	usersMap := make(map[string]UserAuthStruct)
+
+	// isMiddlewarePasswordSet checks if the base is run the first time.
+	err := middleware.checkMiddlewareSetup()
+	if err != nil {
+		return rpcmessages.UserAuthenticateResponse{
+			ErrorResponse: &rpcmessages.ErrorResponse{
+				Success: false,
+				Message: "authentication failed, redis error",
+				Code:    rpcmessages.ErrorAuthenticationFailed,
+			},
+		}
+	}
+
+	if !middleware.isMiddlewarePasswordSet {
+		if args.Username == "admin" && args.Password == initialAdminPassword {
```

Comment:
What is the password is not set (`middleware.isMiddlewarePasswordSet == true`) but the wrong credentials are provided `(args.Username == "admin" && args.Password == initialAdminPassword) == false`?

I don't see an `authentication unsuccessful` error in the top if block to handle this case?

Reply:
Actually that entire if block can go, the rest of the code should handle it correctly. I added it to make it clear that the initial authentication is a special case, but I don't see much point in it any more. Once this is in, I'll do some more research to handle the initial case properly. I removed it now.